File: kotlin/src/main/kotlin/tf/virtual/dojo/algo/core/Actions.kt
package tf.virtual.dojo.algo.core

class Actions {

    private val queue = mutableListOf<String>()

    fun message(message: String) {
        queue.add("MSG $message")
    }

    fun move(source: Factory, destination: Factory, nbCyborgs: Int) {
        queue.add("MOVE ${source.id} ${destination.id} $nbCyborgs")
    }

    fun bomb(source: Factory, destination: Factory) {
        queue.add("BOMB ${source.id} ${destination.id}")
    }

    fun upgrade(factory: Factory) {
        queue.add("INC ${factory.id}")
    }

    fun commit() {
        if (queue.isEmpty()) {
            printDebug("No action found /!\\")
            queue.add("WAIT")
        }
        println(queue.joinToString(";"))
        queue.clear()
    }
}

File: typescript/src/core/actions.ts
import { printDebug } from './lib';
import { Factory } from './models';

class Actions {
    private queue: string[] = [];

    message(message: string) {
        this.queue.push(`MSG ${message}`)
    }

    move(source: Factory, destination: Factory, nbCyborgs: number) {
        this.queue.push(`MOVE ${source.id} ${destination.id} ${nbCyborgs}`)
    }

    bomb(source: Factory, destination: Factory) {
        this.queue.push(`BOMB ${source.id} ${destination.id}`)
    }

    upgrade(factory: Factory) {
        this.queue.push(`INC ${factory.id}`)
    }

    commit() {
        if (this.queue.length == 0) {
            printDebug(`No action found /!\\`);
            this.queue.push('WAIT');
        }
        console.log(this.queue.join(';'));
        this.queue = [];
    }
}

export const actions = new Actions();
